### PR TITLE
Azure auth with SPN now uses AAD token by default instead of PAT

### PR DESCRIPTION
**Pull request by alexott**
* PAT token still could be used by setting `azure_use_pat_for_spn` to `true`
* Check was added to make sure that Azure KeyVault + SPN is detected during plan, not during execution

### CODE REVIEW

File: access/resource_secret_scope.go [hunk 1]
```diff
@@ -150,4 +153,20 @@ func ResourceSecretScope() *schema.Resource {
 			return NewSecretScopesAPI(ctx, c).Delete(d.Id())
 		},
 	}.ToResource()
+
+	resource.CustomizeDiff = customdiff.Sequence(
```
Comment (nfx) [inline access/resource_secret_scope.go:157]: why do we need a sequence for one function?

File: access/resource_secret_scope.go [hunk 2]
```diff
@@ -150,4 +153,20 @@ func ResourceSecretScope() *schema.Resource {
...
+		func(ctx context.Context, diff *schema.ResourceDiff, v interface{}) error {
+			kvLst := diff.Get("keyvault_metadata").([]interface{})
+			if len(kvLst) == 0 {
+				return nil
```
Comment (nfx) [inline access/resource_secret_scope.go:161]: is it still possible to create databricks scopes?

File: common/azure_auth.go [hunk 3]
```diff
@@ -270,6 +277,19 @@ func (aa *AzureAuth) patRequest() tokenRequest {
 	}
 }
 
+func maybeExtendAuthzError(err error) error {
+	var ae APIError
+	fmtString := "Azure authorization error.  Does your SPN  have Contributor access to Databricks workspace. %v"
```
Comment (nfx) [inline common/azure_auth.go:282]: nit: double-spaces 

missing question mark

File: common/azure_auth.go [hunk 4]
```diff
@@ -270,6 +277,19 @@ func (aa *AzureAuth) patRequest() tokenRequest {
...
+	if errors.As(err, &ae) {
+		if ae.ErrorCode == "403" {
```
Comment (nfx) [inline common/azure_auth.go:284]: isn't it `ae.StatusCode == 403` instead?

File: access/resource_secret_scope.go [hunk 5]
```diff
@@ -123,11 +124,13 @@ func ResourceSecretScope() *schema.Resource {
 		s["name"].ValidateFunc = validScope
 		s["initial_manage_principal"].ForceNew = true
 		s["keyvault_metadata"].ForceNew = true
+
 		return s
 	})
-	return common.Resource{
+	resource := common.Resource{
```
Comment (nfx) [inline access/resource_secret_scope.go:130]: probably make CustomizeDiff package-local function and add unit tests for corner cases

File: provider/provider_test.go [hunk 6]
```diff
@@ -350,6 +356,7 @@ func configureProviderAndReturnClient(t *testing.T, tt providerConfigTest) (*com
 	if err != nil {
 		return nil, err
 	}
+	client.AzureAuth.UsePATForSPN = tt.usePATForSPN
```
Comment (nfx) [inline provider/provider_test.go:359]: probably this line should go before `client.Authenticate()`

File: common/azure_auth.go [hunk 7]
```diff
@@ -270,6 +277,19 @@ func (aa *AzureAuth) patRequest() tokenRequest {
...
+			return fmt.Errorf(fmtString, err)
+		}
+	} else if strings.Contains(err.Error(), "does not have authorization to perform action") {
+		return fmt.Errorf(fmtString, err)
```
Comment (nfx) [inline common/azure_auth.go:288]: maybe add unit tests for these two corner cases? :)

File: common/azure_auth.go [hunk 8]
```diff
@@ -270,6 +277,19 @@ func (aa *AzureAuth) patRequest() tokenRequest {
 	}
 }
 
+func maybeExtendAuthzError(err error) error {
+	var ae APIError
+	fmtString := "Azure authorization error.  Does your SPN  have Contributor access to Databricks workspace. %v"
+	if errors.As(err, &ae) {
```
Comment (nfx) [inline common/azure_auth.go:283]: just in case, there might be shorter version:

```go
if e, ok := err.(APIError); ok && e.StatusCode == 403 {
  // ..
}
```

**Review by nfx — CHANGES_REQUESTED** [review]

File: access/resource_secret_scope.go (hunk 2 above)
Comment (alexott) [reply access/resource_secret_scope.go:161]: yes

File: common/azure_auth.go (hunk 4 above)
Comment (alexott) [reply common/azure_auth.go:284]: yeah....

File: access/resource_permissions.go [hunk 9]
```diff
@@ -366,32 +365,30 @@ func ResourcePermissions() *schema.Resource {
 	}
 	return &schema.Resource{
 		Schema: s,
-		CustomizeDiff: customdiff.Sequence(
-			func(ctx context.Context, diff *schema.ResourceDiff, m interface{}) error {
-				me, err := identity.NewUsersAPI(ctx, m).Me()
-				if err != nil {
-					return err
+		CustomizeDiff: func(ctx context.Context, diff *schema.ResourceDiff, m interface{}) error {
```
Comment (nfx) [inline access/resource_permissions.go:368]: can you revert the formatting for this file not to interfere with diff?

File: access/resource_secret_scope.go [hunk 10]
```diff
@@ -114,6 +114,20 @@ var validScope = validation.StringMatch(regexp.MustCompile(`^[\w\.@_-]{1,128}$`)
 	"Must consist of alphanumeric characters, dashes, underscores, and periods, "+
 		"and may not exceed 128 characters.")
 
+func kvDiffFunc(ctx context.Context, diff *schema.ResourceDiff, v interface{}) error {
+	if diff != nil {
```
Comment (nfx) [inline access/resource_secret_scope.go:118]: `if diff == nil { return nil }` would be better

File: access/resource_secret_scope_test.go [hunk 11]
```diff
@@ -301,3 +302,39 @@ func TestResourceSecretScopeDelete_Error(t *testing.T) {
...
+		HCL: `
+			name = "Boom"
+			keyvault_metadata {
+				resource_id = "bcd"
+				dns_name = "def"
+			}`,
+		AzureAuth: &common.AzureAuth{ClientID: "123", ClientSecret: "123", TenantID: "123",
+			ResourceID: "/subscriptions/a/resourceGroups/b/providers/Microsoft.Databricks/workspaces/c"},
+		Create: true,
+	}.Apply(t)
```
[the remark this answers is not in the document]
Comment (nfx) [reply access/resource_secret_scope_test.go:337]: you can also use `qa.HTTPFixturesApply` if you want to test just the function and not do the whole fixture

File: provider/provider.go [hunk 12]
```diff
@@ -211,6 +211,12 @@ func DatabricksProvider() *schema.Provider {
 				Default:     false,
 				Description: "Create ephemeral PAT tokens also for AZ CLI authenticated requests",
 			},
+			"azure_use_pat_for_spn": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
```
Comment (nfx) [inline provider/provider.go:217]: could you, just in case, make it as EnvDefaultFunc?

**Review by nfx — CHANGES_REQUESTED** [review]
last round of small comments

File: access/resource_permissions.go (hunk 9 above)
Comment (alexott) [reply access/resource_permissions.go:368]: it's not a pure formatting, it's removal of `customdiff.Sequence`, just applying the earlier comment to multiple places.  But I can remove it

File: qa/testing.go [hunk 13]
```diff
@@ -102,6 +103,9 @@ func (f ResourceFixture) Apply(t *testing.T) (*schema.ResourceData, error) {
 	if f.Azure {
 		client.AzureAuth.ResourceID = "/subscriptions/a/resourceGroups/b/providers/Microsoft.Databricks/workspaces/c"
 	}
+	if f.AzureAuth != nil {
+		client.AzureAuth = *f.AzureAuth
```
[the remark this answers is not in the document]
Comment (alexott) [reply qa/testing.go:107]: I didn't want to make it more complex than necessary - with `f.Azure` it's easy, just set it to `true`, and it's all. with `AzureAuth` it's slightly more complex

**Review by nfx — APPROVED** [review]
